Changes in module 'nf-core/eggnogmapper'
--- modules/nf-core/eggnogmapper/main.nf
+++ modules/nf-core/eggnogmapper/main.nf
@@ -1,6 +1,6 @@
 process EGGNOGMAPPER {
     tag "$meta.id"
-    label 'process_long'
+    label 'process_high'
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
@@ -9,9 +9,11 @@
 
     input:
     tuple val(meta), path(fasta)
-    path(eggnog_db)
     path(eggnog_data_dir)
-    tuple val(meta2), path(eggnog_diamond_db)
+    path(db)
+    path(proteins_dmnd)
+    path(taxa_db)
+    path(taxa_db_pkl)
 
     output:
     tuple val(meta), path("*.emapper.annotations")   , emit: annotations
@@ -37,9 +39,7 @@
         --cpu ${task.cpus} \\
         -i ${fasta_name} \\
         --data_dir ${eggnog_data_dir} \\
-        -m diamond \\
-        --dmnd_db ${eggnog_diamond_db} \\
-        --database ${eggnog_db} \\
+        -m diamond --tax_scope prokaryota_broad --pfam_realign realign \\
         --output ${prefix} \\
         ${dbmem} \\
         $args

************************************************************
